src/http.cc

```diff
@@ -829,11 +841,142 @@ HttpStateData::handle1xx(HttpReply *reply)
     // for similar reasons without a 1xx response.
 }
 
+/// A protocol-comparing predicate for STL search algorithms.
+/// Helps to match a versioned reference protocol name to a
+/// name without version (for example the reference 'websocket/1.2'
+/// should match to 'websocket').
+class MatchProtocol
+{
+public:
+    explicit MatchProtocol(const char *s, size_t len) : reference(s), referenceLen(len) {
+        const char *end = std::find(reference, reference + referenceLen, '/');
+        referenceBaseLen = end - reference;
+    }
+
+    /// Operator to search in lists or vectors
+    bool operator() (const SBuf &checking) {
+        if (referenceIsVersioned()) {
+            const bool checkingIsVersioned = (checking.find('/') != SBuf::npos);
+            if (!checkingIsVersioned) {
+                return checking.length() == referenceBaseLen && checking.caseCmp(reference, referenceBaseLen) == 0;
+            }
+        }
+        return checking.length() == referenceLen && checking.caseCmp(reference, referenceLen) == 0;
+    }
+
+    /// Operator to search in std::maps or similar
+    template<typename T>
+    bool operator() (const std::pair<const SBuf, T> &checkingPair) {
+        return (*this)(checkingPair.first);
+    }
+
+private:
+    /// \return true if the reference protocol includes version info
+    bool referenceIsVersioned() { return (referenceBaseLen != referenceLen); }
+
+    const char *reference; ///< The reference protocol
+
+    /// The full length of reference protocol string, including the
+    /// version part
+    size_t referenceLen;
+
+    /// The length of reference protocol string without version part
+    size_t referenceBaseLen;
+};
+
+bool
+HttpStateData::upgradeProtocolsSupported(const HttpReply *reply) const
+{
+    const String upgradeProtos = reply->header.getList(Http::HdrType::UPGRADE);
+
+    if (!upgradeHeaderOut) {
+        debugs(11, 2, "Upgrade to '" << upgradeProtos << "' is not requested");
+        return false;
+    }
+
+    const char *pos = nullptr;
+    const char *item;
+    int ilen;
+    while (strListGetItem(&upgradeProtos, ',', &item, &ilen, &pos)) {
+        const auto *e = std::find(item, item + ilen, '/');
+        ilen = e - item; //Fix ilen to not include protocol version
```

Comment:
If I understand your intention correctly, we now match only protocol names here, ignoring versions. Is it correct? If it is so, we should probably add a comment, briefly explaining this decision, based probably on the related discussion [thread](https://github.com/measurement-factory/squid/pull/16#discussion_r242654279).

Reply:
Yes it is correct.
My sense is that explaining the decision can go to a wiki page. 
Also there is related documentation in cf.data.pre:
>     By default, Squid does not restrict Upgrade response protocols because
>      it is often difficult for a proxy to know for sure whether a combination
>       of protocols and their versions is meaningful to the client. Squid
>       behavior may become more restrictive in the future. Meanwhile, an admin
>       can restrict HTTP 101 (Switching Protocols) responses using
>       http_reply_access. Denied 101 responses result in XXX errors and server
>       connection closures.

I do not know if it make sense to include the first phrase from the above.
My opinion is that it is not required.